feat(robot-server): create middleware to support api versioning

amitlissack opened the pull request:
## overview

Add API version header to responses.

## changelog

Define an api version. This is an integer starting at 1.

Middleware that examines the `opentrons-version` header in request.  The api version used to respond to the request is chosen in this manner: 
- if the request version is an integer and <= server api version, we use that version
- otherwise we use the server version.

The chosen version is attached to the request.state object for consumption by handlers.
Finally, the response will contain `opentrons-version` that holds the version used to handle the request.

Rename `exceptions.py` to `errors.py` and put error related helpers in there.
 
## review requests

- [x] Version string is ok
- [x] Header name is ok
- [x] Location of api version constant is ok
- [x] This closes #4633. (The ticket discusses validating request version headers. Is that something we want to do now?)

## risk assessment

None


closes #4633

mcous commented on robot-server/robot_server/__init__.py:
```diff
@@ -0,0 +1,3 @@
+# This is the version of the API supported by the server. The
+# application version is in package.json.
+API_VERSION = "1.0.0"
```
As per #4633, this should be a simple integer that counts up. Any breaking change (major bump) should involve path changes (e.g. `/v2/`), but hopefully we can avoid needing to do that for a long time

mcous commented on robot-server/robot_server/service/app.py:
```diff
@@ -57,11 +57,24 @@ async def on_startup():
 
 
+RESPONSE_HEADERS = {
+    'X-Opentrons-Media-Type': f'opentrons.robot-server.{API_VERSION}'
```
I updated the #4633 ticket with a few thoughts on this:

- `X-...` media types [have been deprecated](https://tools.ietf.org/html/rfc6648) in favor of naming headers in a way to inherintly avoid collision (e.g. prefixing all our custom headers with `Opentrons-`)
- String parsing is a pain for this
- I'm down to go something way more simple, like `'Opentrons-Version': API_VERSION`

amitlissack commented:
> Sorry for not doing this sooner, but I just went in and updated the #4633 spec. I simplified the spec based pretty much entirely on what the [Stripe API does](https://stripe.com/docs/api) because I like that API a lot.
> 
> Also, #4633 contains two halves: requests and response. Since this PR only addresses the response half, *_it does not close the ticket as written_. We should either keep the ticket open or write a new one for the request side of things

I think a separate ticket makes more sense.

sfoster1 commented on robot-server/robot_server/service/errors.py:
```diff
@@ -0,0 +1,77 @@
+    Default schema of fastAPI exception response is:
+    {
+        'loc': ('body',
+                '<outer_scope1>',
```
is this precisely 2 outer scopes or is it (body, [0-n outer scopes], inner_param)

amitlissack replied:
It is the latter.
